dialogue-core/src/main/java/com/palantir/dialogue/core/RetryOtherValidatingChannel.java

```diff
@@ -0,0 +1,131 @@
+import com.google.common.net.HttpHeaders;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.RateLimiter;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class RetryOtherValidatingChannel implements Channel {
+
+    private static final Logger log = LoggerFactory.getLogger(RetryOtherValidatingChannel.class);
+    private static final RateLimiter VALIDATION_FAILED_LOGGING_LIMITER = RateLimiter.create(1);
+
+    private final Channel delegate;
+    private final List<String> hosts;
+    private final Consumer<String> failureReporter;
+
+    RetryOtherValidatingChannel(Channel delegate, List<String> hosts) {
+        this(delegate, hosts, RetryOtherValidatingChannel.failureReporter(hosts));
+    }
+
+    @VisibleForTesting
+    RetryOtherValidatingChannel(Channel delegate, List<String> hosts, Consumer<String> failureReporter) {
+        this.delegate = delegate;
+        this.hosts =
+                hosts.stream().map(RetryOtherValidatingChannel::strictParseHost).collect(Collectors.toList());
+        this.failureReporter = failureReporter;
+    }
+
+    @Override
+    @SuppressWarnings("Finally")
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return DialogueFutures.transform(delegate.execute(endpoint, request), response -> {
+            try {
+                validateRetryOther(response);
+                return response;
+            } finally {
+                return response;
+            }
+        });
+    }
+
+    private void validateRetryOther(Response response) {
+        if (!Responses.isRetryOther(response)) {
+            return;
+        }
+        Optional<String> maybeRetryOtherUri = response.getFirstHeader(HttpHeaders.LOCATION);
+        if (maybeRetryOtherUri.isPresent()) {
+            String retryOtherUri = maybeRetryOtherUri.get();
+            if (!isValidUri(retryOtherUri)) {
+                failureReporter.accept(retryOtherUri);
+            }
+        }
+    }
+
+    private boolean isValidUri(String uri) {
+        String maybeHost = maybeParseHost(uri);
+        return (maybeHost != null) && isKnownHost(maybeHost);
+    }
+
+    private boolean isKnownHost(String host) {
+        for (int i = 0; i < hosts.size(); i++) {
+            if (hosts.get(i).equals(host)) {
+                return true;
+            }
+        }
+        return false;
```

Comment:
If we use a set, this can be simplified to `hosts.contains(host)`

Reply:
Done